File: services/identity-service/src/main/java/br/com/deltaglobalbank/identity/domain/apiKey/ApiKey.java
package br.com.deltaglobalbank.identity.domain.apiKey;

import java.time.Instant;
import java.util.Objects;
import java.util.UUID;

import br.com.deltaglobalbank.identity.domain.user.HashedPassword;

public class ApiKey {

    public static final int MIN_NAME_LENGTH = 3;
    public static final int MAX_NAME_LENGTH = 255;

    private final UUID id;
    private final UUID apiClientId;
    private final String name;
    private final HashedPassword keyHash;
    private final String keyPrefix;
    private final String fingerprint;
    private final Instant createdAt;

    private Instant expiresAt;
    private Instant lastUsedAt;
    private Instant revokedAt;
    private Instant updatedAt;

    private ApiKey(
        UUID id,
        UUID apiClientId,
        String name,
        HashedPassword keyHash,
        String keyPrefix,
        String fingerprint,
        Instant expiresAt,
        Instant lastUsedAt,
        Instant revokedAt,
        Instant createdAt,
        Instant updatedAt
    ) {
        this.id = id;
        this.apiClientId = apiClientId;
        this.name = name;
        this.keyHash = keyHash;
        this.keyPrefix = keyPrefix;
        this.fingerprint = fingerprint;
        this.expiresAt = expiresAt;
        this.lastUsedAt = lastUsedAt;
        this.revokedAt = revokedAt;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
    }

    public UUID getId() {
        return id;
    }

    public UUID getApiClientId() {
        return apiClientId;
    }

    public String getName() {
        return name;
    }

    public HashedPassword getKeyHash() {
        return keyHash;
    }

    public String getKeyPrefix() {
        return keyPrefix;
    }

    public String getFingerprint() {
        return fingerprint;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public boolean isRevoked() {
        return revokedAt != null;
    }

    public boolean isExpired() {
        return expiresAt != null && expiresAt.isBefore(Instant.now());
    }

    public boolean isActive() {
        return !isRevoked() && !isExpired();
    }

    public void revoke() {
        if (isRevoked()) {
            return;
        }
        Instant now = Instant.now();
        revokedAt = now;
        updatedAt = now;
    }

    public void markUsed() {
        Instant now = Instant.now();
        lastUsedAt = now;
        updatedAt = now;
    }

    public Snapshot snapshot() {
        return new Snapshot(
            expiresAt,
            lastUsedAt,
            revokedAt,
            updatedAt
        );
    }

    public record Snapshot(
        Instant expiresAt,
        Instant lastUsedAt,
        Instant revokedAt,
        Instant updatedAt
    ) {
    }

    @Override
    public boolean equals(Object other) {
        return other instanceof ApiKey apiKey && apiKey.id.equals(id);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }

    public static ApiKey create(
        UUID id,
        UUID apiClientId,
        String name,
        HashedPassword keyHash,
        String keyPrefix,
        String fingerprint,
        Instant expiresAt
    ) {
        if (name.length() < MIN_NAME_LENGTH || name.length() > MAX_NAME_LENGTH) {
            throw new IllegalArgumentException("invalid_api_key_name_length");
        }
        if (keyPrefix.isBlank()) {
            throw new IllegalArgumentException("invalid_key_prefix");
        }
        if (fingerprint.length() != 64) {
            throw new IllegalArgumentException("invalid_fingerprint");
        }
        if (expiresAt != null && !expiresAt.isAfter(Instant.now())) {
            throw new IllegalArgumentException("expires_at_in_past");
        }

        Instant now = Instant.now();
        return new ApiKey(
            id,
            apiClientId,
            name.trim(),
            keyHash,
            keyPrefix,
            fingerprint,
            expiresAt,
            null,
            null,
            now,
            now
        );
    }

    public static ApiKey restore(
        UUID id,
        UUID apiClientId,
        String name,
        HashedPassword keyHash,
        String keyPrefix,
        String fingerprint,
        Instant expiresAt,
        Instant lastUsedAt,
        Instant revokedAt,
        Instant createdAt,
        Instant updatedAt
    ) {
        return new ApiKey(
            id,
            apiClientId,
            name,
            keyHash,
            keyPrefix,
            fingerprint,
            expiresAt,
            lastUsedAt,
            revokedAt,
            createdAt,
            updatedAt
        );
    }
}

File: services/identity-service/src/main/kotlin/br/com/deltaglobalbank/identity/domain/apiKey/ApiKey.kt
package br.com.deltaglobalbank.identity.domain.apiKey

import br.com.deltaglobalbank.identity.domain.user.HashedPassword
import java.time.Instant
import java.util.UUID

class ApiKey private constructor(
    val id: UUID,
    val apiClientId: UUID,
    val name: String,
    val keyHash: HashedPassword,
    val keyPrefix: String,
    val fingerprint: String,
    private var _expiresAt: Instant?,
    private var _lastUsedAt: Instant?,
    private var _revokedAt: Instant?,
    val createdAt: Instant,
    private var _updatedAt: Instant
) {

    fun isRevoked(): Boolean = _revokedAt != null
    fun isExpired(): Boolean = _expiresAt?.isBefore(Instant.now()) ?: false
    fun isActive(): Boolean = !isRevoked() && !isExpired()

    fun revoke() {
        if (isRevoked()) return  // idempotente
        val now = Instant.now()
        _revokedAt = now
        _updatedAt = now
    }

    fun markUsed() {
        val now = Instant.now()
        _lastUsedAt = now
        _updatedAt = now
    }

    fun snapshot(): Snapshot = Snapshot(
        expiresAt = _expiresAt,
        lastUsedAt = _lastUsedAt,
        revokedAt = _revokedAt,
        updatedAt = _updatedAt
    )

    data class Snapshot(
        val expiresAt: Instant?,
        val lastUsedAt: Instant?,
        val revokedAt: Instant?,
        val updatedAt: Instant
    )

    override fun equals(other: Any?): Boolean = other is ApiKey && other.id == this.id
    override fun hashCode(): Int = id.hashCode()

    companion object {
        const val MIN_NAME_LENGTH = 3
        const val MAX_NAME_LENGTH = 255

        fun create(
            id: UUID,
            apiClientId: UUID,
            name: String,
            keyHash: HashedPassword,
            keyPrefix: String,
            fingerprint: String,
            expiresAt: Instant?
        ): ApiKey {
            require(name.length in MIN_NAME_LENGTH..MAX_NAME_LENGTH) {
                "invalid_api_key_name_length"
            }
            require(keyPrefix.isNotBlank()) { "invalid_key_prefix" }
            require(fingerprint.length == 64) { "invalid_fingerprint" }
            expiresAt?.let {
                require(it.isAfter(Instant.now())) { "expires_at_in_past" }
            }

            val now = Instant.now()
            return ApiKey(
                id = id,
                apiClientId = apiClientId,
                name = name.trim(),
                keyHash = keyHash,
                keyPrefix = keyPrefix,
                fingerprint = fingerprint,
                _expiresAt = expiresAt,
                _lastUsedAt = null,
                _revokedAt = null,
                createdAt = now,
                _updatedAt = now
            )
        }

        fun restore(
            id: UUID,
            apiClientId: UUID,
            name: String,
            keyHash: HashedPassword,
            keyPrefix: String,
            fingerprint: String,
            expiresAt: Instant?,
            lastUsedAt: Instant?,
            revokedAt: Instant?,
            createdAt: Instant,
            updatedAt: Instant
        ): ApiKey = ApiKey(
            id = id,
            apiClientId = apiClientId,
            name = name,
            keyHash = keyHash,
            keyPrefix = keyPrefix,
            fingerprint = fingerprint,
            _expiresAt = expiresAt,
            _lastUsedAt = lastUsedAt,
            _revokedAt = revokedAt,
            createdAt = createdAt,
            _updatedAt = updatedAt
        )
    }
}